File: fuzz/corpus/js/vulnerable.js
// Zentinel ground truth — every line with a // TRIGGER comment should fire the named rule.
// Lines without TRIGGER should NOT fire any rule.
const child_process = require('child_process');
const crypto = require('crypto');
const http = require('http');

// ── Code Injection ──────────────────────────────────────
function evalInput(code) {
    eval(code);                                     // TRIGGER javascript.security.eval-usage
}

function funcConstructor(code) {
    return Function(code);                          // TRIGGER javascript.security.function-constructor
}

function delayedEval(code) {
    setTimeout(code, 1000);                         // TRIGGER javascript.security.settimeout-string
}

function repeatedEval(code) {
    setInterval(code, 1000);                        // TRIGGER javascript.security.setinterval-string
}

// ── Command Injection ───────────────────────────────────
function runCommand(cmd) {
    exec(cmd);                                      // TRIGGER javascript.security.exec-usage
}

function cpExec(cmd) {
    child_process.exec(cmd);                        // TRIGGER javascript.security.child-process-exec
}

function cpSpawn(cmd) {
    child_process.spawn(cmd);                       // TRIGGER javascript.security.child-process-spawn
}

// ── XSS / DOM ───────────────────────────────────────────
function writeHtml(html) {
    document.write(html);                           // TRIGGER javascript.security.innerhtml
}

// ── Deserialization ─────────────────────────────────────
function parseData(data) {
    return JSON.parse(data);                        // TRIGGER javascript.security.json-parse
}

// ── Cryptography ────────────────────────────────────────
function hashData(data) {
    return crypto.createHash("md5");                // TRIGGER javascript.security.crypto-createhash-md5
}

// ── Network / TLS ───────────────────────────────────────
function startServer() {
    http.createServer(handler);                     // TRIGGER javascript.security.http-createserver
}

// ── Hardcoded Secrets ───────────────────────────────────
const API_KEY = "sk-1234567890abcdef";              // TRIGGER javascript.security.hardcoded-secret
const DB_PASSWORD = "hunter2";                      // TRIGGER javascript.security.hardcoded-secret

// ── Dangerous Patterns ──────────────────────────────────
function shutdown() {
    process.exit(1);                                // TRIGGER javascript.security.process-exit
}

File: test_fixtures/python/vulnerable.py
# Zentinel ground truth — every line with a # TRIGGER comment should fire the named rule.
# Lines without TRIGGER should NOT fire any rule.
import os
import subprocess
import pickle
import yaml
import marshal
import shelve
import hashlib
import ssl
import requests
import tempfile

# ── Command Injection ────────────────────────────────────
def run_user_input(cmd):
    exec(cmd)                                       # TRIGGER python.security.exec-usage

def evaluate_expression(expr):
    return eval(expr)                               # TRIGGER python.security.eval-usage

def system_call(cmd):
    os.system(cmd)                                  # TRIGGER python.security.os-system

def popen_call(cmd):
    os.popen(cmd)                                   # TRIGGER python.security.os-popen

def subprocess_call(cmd):
    subprocess.call(cmd, shell=True)                # TRIGGER python.security.subprocess-shell

def subprocess_run_cmd(cmd):
    subprocess.run(cmd, shell=True)                 # TRIGGER python.security.subprocess-run

def subprocess_popen_cmd(cmd):
    subprocess.Popen(cmd, shell=True)               # TRIGGER python.security.subprocess-popen

# ── Deserialization ──────────────────────────────────────
def load_pickle(f):
    return pickle.load(f)                           # TRIGGER python.security.pickle-load

def loads_pickle(data):
    return pickle.loads(data)                       # TRIGGER python.security.pickle-loads

def load_yaml(f):
    return yaml.load(f)                             # TRIGGER python.security.yaml-load

def load_marshal(data):
    return marshal.loads(data)                      # TRIGGER python.security.marshal-loads

def open_shelve(path):
    return shelve.open(path)                        # TRIGGER python.security.shelve-open

# ── Cryptography ─────────────────────────────────────────
def hash_md5(data):
    return hashlib.md5(data)                        # TRIGGER python.security.hashlib-md5

def hash_sha1(data):
    return hashlib.sha1(data)                       # TRIGGER python.security.hashlib-sha1

# ── Dangerous Builtins ───────────────────────────────────
def compile_code(source):
    return compile(source, "<string>", "exec")      # TRIGGER python.security.compile-usage

def get_input():
    return input("Enter value: ")                   # TRIGGER python.security.input-python2

# ── Network / TLS ────────────────────────────────────────
def no_verify_ssl():
    return ssl._create_unverified_context()         # TRIGGER python.security.ssl-no-verify

def fetch_url(url):
    return requests.get(url, verify=False)          # TRIGGER python.security.requests-no-verify

# ── Temp Files ───────────────────────────────────────────
def make_temp():
    return tempfile.mktemp()                        # TRIGGER python.security.tempfile-mktemp

# ── Tier 1: Precise Weak Cryptography ────────────────────
def hash_new_md5(data):
    return hashlib.new("md5", data)                 # TRIGGER python.security.hashlib-new-md5

def hash_new_sha1(data):
    return hashlib.new("sha1", data)                # TRIGGER python.security.hashlib-new-sha1

# ── Tier 1: Code Injection via F-String ──────────────────
def eval_fstring(user):
    eval(f"print({user})")                          # TRIGGER python.security.eval-fstring (+ eval-usage)

def exec_fstring(code):
    exec(f"result = {code}")                        # TRIGGER python.security.exec-fstring (+ exec-usage)

# ── Hardcoded Secrets ────────────────────────────────────
API_KEY = "sk-1234567890abcdef"                     # TRIGGER python.security.hardcoded-secret
DATABASE_PASSWORD = "hunter2"                       # TRIGGER python.security.hardcoded-secret
